src/Features/Core/Portable/Highlighting/Keywords/AbstractKeywordHighlighter.cs

```diff
@@ -2,54 +2,61 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#nullable disable
-
 using System.Collections.Generic;
 using System.Threading;
 using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Collections;
 using Microsoft.CodeAnalysis.Text;
 
 namespace Microsoft.CodeAnalysis.Highlighting;
 
-internal abstract class AbstractKeywordHighlighter<TNode> : AbstractKeywordHighlighter where TNode : SyntaxNode
+internal abstract class AbstractKeywordHighlighter<TNode>(bool findInsideTrivia = true)
+    : AbstractKeywordHighlighter(findInsideTrivia)
+    where TNode : SyntaxNode
 {
-    protected sealed override bool IsHighlightableNode(SyntaxNode node) => node is TNode;
+    protected sealed override bool IsHighlightableNode(SyntaxNode node)
+        => node is TNode;
 
     protected sealed override void AddHighlightsForNode(SyntaxNode node, List<TextSpan> highlights, CancellationToken cancellationToken)
         => AddHighlights((TNode)node, highlights, cancellationToken);
 
     protected abstract void AddHighlights(TNode node, List<TextSpan> highlights, CancellationToken cancellationToken);
 }
 
-internal abstract class AbstractKeywordHighlighter : IHighlighter
+internal abstract class AbstractKeywordHighlighter(bool findInsideTrivia = true) : IHighlighter
 {
+    private readonly bool _findInsideTrivia = findInsideTrivia;
+
     private static readonly ObjectPool<List<TextSpan>> s_textSpanListPool = new(() => []);
-    private static readonly ObjectPool<List<SyntaxToken>> s_tokenListPool = new(() => []);
 
     protected abstract bool IsHighlightableNode(SyntaxNode node);
 
+    protected virtual bool ContainsHighlightableToken(ref TemporaryArray<SyntaxToken> tokens)
+        => true;
+
     public void AddHighlights(
         SyntaxNode root, int position, List<TextSpan> highlights, CancellationToken cancellationToken)
     {
-        using (s_textSpanListPool.GetPooledObject(out var tempHighlights))
-        using (s_tokenListPool.GetPooledObject(out var touchingTokens))
-        {
-            AddTouchingTokens(root, position, touchingTokens);
+        // We only look at a max of 4 tokens (two trivia, and two non-trivia), so a temp-array is ideal here
+        using var touchingTokens = TemporaryArray<SyntaxToken>.Empty;
+        AddTouchingTokens(root, position, ref touchingTokens.AsRef());
+        if (!ContainsHighlightableToken(ref touchingTokens.AsRef()))
```

Comment:
the async/await highlighter si not well written.  it does all sorts of unnecessary work even when not on an appropraite keyword.  this allows us to early bail out here and not do any of that work.